onmt/data/SampledDataset.lua

```diff
@@ -0,0 +1,266 @@
+--[[ Data management and batch creation. Handles data created by `preprocess.lua`. ]]
+require 'torchx'
+
+local SampledDataset = torch.class("SampledDataset")
```

Comment:
I reverted back to preparing batch by iterating over all instances due to batch efficiency and to keep correct attention weights without using maskedsoftmax.

However, to allow oversampling, Instead of adding one instance when it is marked as sampled, now instances can be added multiple time.

This will potentially bias the batch towards more frequently sampled ones and will have more effect (whether positive or negative) on training than before.